polygeist/tools/cgeist/Lib/ValueCategory.h

```diff
@@ -9,15 +9,25 @@
 #ifndef CLANG_MLIR_VALUE_CATEGORY
 #define CLANG_MLIR_VALUE_CATEGORY
 
+#include "mlir/Dialect/Arith/IR/Arith.h"
 #include "mlir/IR/Builders.h"
 #include "mlir/IR/Value.h"
+#include "llvm/Support/WithColor.h"
 
 // Represents a rhs or lhs value.
 class ValueCategory {
 public:
   mlir::Value val;
   bool isReference;
 
+  template <typename OpTy>
+  ValueCategory Cast(mlir::OpBuilder &builder, mlir::Type PromotionType) const {
+    if (val.getType() == PromotionType)
+      return *this;
+    return {builder.createOrFold<OpTy>(val.getLoc(), PromotionType, val),
+            false};
```

Comment:
Although casting of an FP constant is not folded yet, it might in the future. That's why we do not use `create`.